Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoASA.yaml

```diff
@@ -0,0 +1,423 @@
+                | parse Message with * ":  " * " " * " " * " connection " NetworkSessionId " for " SrcInterfaceName ":" * "/" * " (" SrcNatIpAddr "/" SrcNatPortNumber:int ")" SrcUsername "to " DstInterfaceName ":" * "/" * " (" DstNatIpAddr "/" DstNatPortNumber:int ")" DstUsername
+                | extend SrcUsername = trim(@"\s?\(?\)?\s?", SrcUsername),
+                        DstUsername = trim(@"\s?\(?\)?\s?", DstUsername)
+                | extend 
+                        SimplifiedDeviceAction = AliasDeviceAction,
+                        SessionId = NetworkSessionId;
+      let all_302013_302015_unparsed = unparsedData
+                | where AliasDeviceEventClassID in ("302013","302015")
+                | parse Message with * ":  " DeviceAction " " NetworkDirection " " Protocol " connection " NetworkSessionId " for " SrcInterfaceName ":" SrcIpAddr "/" SrcPortNumber:int " (" SrcNatIpAddr "/" SrcNatPortNumber:int ")" SrcUsername "to " DstInterfaceName ":" DstIpAddr "/" DstPortNumber:int " (" DstNatIpAddr "/" DstNatPortNumber:int ")" DstUsername
+                | extend SrcUsername = trim(@"\s?\(?\)?\s?", SrcUsername),
+                        DstUsername = trim(@"\s?\(?\)?\s?", DstUsername),
+                        NetworkDirection = case(NetworkDirection == "inbound", "Inbound",
+                                                    NetworkDirection == "outbound", "Outbound",
+                                                    "")
+                | extend 
+                        SimplifiedDeviceAction = DeviceAction,
+                        SessionId = NetworkSessionId;
+      let all_302014_unparsed = unparsedData
+                | where AliasDeviceEventClassID == "302014"
+                | parse Message with * ": " DeviceAction " " Protocol " connection " NetworkSessionId " for " SrcInfoString " to " DstInfoString " duration " NetworkDuration " bytes " DstBytes:long *
+                | parse kind=regex SrcInfoString with SrcInterfaceName ":" SrcIpAddr "/" SrcPortNumber:int @"\(?\s?" SrcUsername @"\)?\s?"
+                | parse kind=regex DstInfoString with DstInterfaceName ":" DstIpAddr "/" DstPortNumber:int @"\(?\s?" DstUsername @"\)?\s?"
+                | parse Message with * " bytes " * " " RemainingString
+                | parse RemainingString with ReasonString "(" DstUsernameSimple ")"
+                | extend ReasonString = case(RemainingString has "from" and RemainingString !has "(", RemainingString,
+                                             ReasonString)
+                | parse ReasonString with Reason " from " *
+                | extend Reason = case(isempty(Reason), ReasonString,
+                                       Reason)
+                | lookup EventResultMapping on Reason
+                | extend 
+                            SrcUsername = trim(@"\s?\(?\)?\s?", SrcUsername),
+                            DstUsername = case(isempty(DstUsername),DstUsernameSimple,
```

Comment:
I think that there is a missing element in this case statement.

Reply:
There are two types of events here:
1. "%ASA-6-302014:  Teardown TCP connection 261246338 for OUTSIDE:89.160.20.156/50120(LOCAL\\domain\\USER001) to OUTSIDE:40.0.0.1/443 duration 0:02:05 bytes 9610 TCP FINs (domain\\USER001)"
2. "%ASA-6-302014:  Teardown TCP connection 261246338 for OUTSIDE:89.160.20.156/50120(LOCAL\\domain\\USER001) to OUTSIDE:40.0.0.1/443(LOCAL\\domain\\USER002) duration 0:02:05 bytes 9610 TCP FINs (domain\\USER002)"

In case there is no Destination Username after DstPortNumber (DstUsername in code), we take information from the user details (DstUsernameSimple in code) present at the end. Otherwise, we capture only the earlier one.